### PR TITLE
MGDOBR-462: Bridge/ProcessorServices: Improve split of create/delete methods X(..) and doX(..)

### PR DESCRIPTION
**Pull request by manstis**
[https://issues.redhat.com/browse/MGDOBR-462](https://issues.redhat.com/browse/MGDOBR-462)

## PR owner checklist

Please make sure that your PR meets the following requirements:

- [x] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config).  
  *Run `mvn clean verify -DskipTests` so that imports are correctly set.*
- [x] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-1] - $clear_explanation_of_what_you_did"
- [x] All new functionality is tested
- [x] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

## PR reviewer(s) checklist

- [ ] Check if code and Github action workflows have been modified and if the modification is safe
- [ ] If the modification is safe, add the `safe to test` label

<details>
<summary>
How to trigger pipelines and use the bots:
</summary>

- <b>Run the end to end pipeline</b>  
  Annotate the pull request with the label: `safe to test`. If you want to run the pipeline again, remove and add it again.

- <b>Rebase the pull request</b>  
  Comment with: `/rebase`.

</details>

### CODE REVIEW

File: manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java [hunk 1]
```diff
@@ -81,11 +70,20 @@ protected Bridge doCreateBridge(String customerId, BridgeRequest bridgeRequest)
...
-        bridgeDAO.persist(bridge);
+
+        doCreateBridge(bridge);
 
         return bridge;
     }
 
+    @Transactional
+    // Bridge and Work creation should always be in the same transaction
+    protected void doCreateBridge(Bridge bridge) {
```
Comment (robpblake) [inline manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java:81]: @manstis I personally don't see the value of us keeping the `doCreateBridge` any more. We can just hoist the calls here into `createBridge`, add `@Transactional` to `createBridge` and then delete this method IMHO.

File: manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java [hunk 2]
```diff
@@ -121,24 +119,22 @@ public Bridge getBridge(String id, String customerId) {
 
     @Override
     public void deleteBridge(String id, String customerId) {
```
Comment (robpblake) [inline manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java:121]: Same comment as the need to maintain to `doCreateBridge` method.

File: manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java [hunk 3]
```diff
@@ -121,23 +121,22 @@ public Processor createProcessor(String bridgeId, String customerId, ProcessorRe
...
         createProcessorConnectorEntity(newProcessor, actionProvider, resolvedAction);
 
-        workManager.schedule(newProcessor);
-
         return newProcessor;
     }
 
     @Transactional
-    // Connector should always be marked for creation in the same transaction as a Processor
+    // Processor, Connector and Work should always be created in the same transaction
```
Comment (robpblake) [inline manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java:130]: Same general concepts here about method consolidation as in `BridgeServiceImpl` 👍

File: manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java (hunk 1 above)
Comment (manstis) [reply manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java:81]: @robpblake I was being conscious to make the transactional code as short as possible to avoid optimistic lock contention. However since I guess a single record will only ever be updated by either the `manager` or `shard` we're unlikely to ever have contention as both operate on the same record at quite different times.

Comment (manstis) [reply manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java:81]: I'll make the change next week.

Comment (manstis) [reply manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java]: @robpblake Updates completed.

File: manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java (hunk 2 above)
Comment (manstis) [reply manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java:119]: @robpblake Updates completed.

File: manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java (hunk 3 above)
Comment (manstis) [reply manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java]: @robpblake Updates completed.

File: manager/src/test/java/com/redhat/service/bridge/manager/workers/resources/ConnectorWorkerTest.java [hunk 4]
```diff
@@ -88,6 +89,7 @@ void handleWorkProvisioningWithUnknownResource() {
         assertThatCode(() -> worker.handleWork(work)).isInstanceOf(IllegalStateException.class);
     }
 
+    @Transactional
```
Comment (lucamolteni) [inline manager/src/test/java/com/redhat/service/bridge/manager/workers/resources/ConnectorWorkerTest.java:92]: I wonder why is this needed now?

**Review by lucamolteni — APPROVED** [review]

File: manager/src/test/java/com/redhat/service/bridge/manager/workers/resources/ConnectorWorkerTest.java (hunk 4 above)
Comment (manstis) [reply manager/src/test/java/com/redhat/service/bridge/manager/workers/resources/ConnectorWorkerTest.java:92]: Because I've made `WorkManager.schedule(..)` demand a transaction (as work should only be scheduled in the same transaction as the resource is created.. failure to create the resource or work, for whatever reason, should rollback both). The tests use the DAO directly (not services to create/delete bridges/processors) so need now to become transactional for their use of `WorkManger.schedule(..)` to function.

**Review by robpblake — APPROVED** [review]

**Comment by manstis** [comment]
Hello @jomarko could you please take a look.

**Comment by jomarko** [comment]
I miss integration tests results in CI
Screenshot from 2022-03-29 13-04-18

**Comment by jomarko** [comment]
Is it ok `Build all images and run E2E tests` was not started?

**Comment by manstis** [comment]
Hi @jomarko. I believe there is an issue with the ITs somewhere (not sure what, perhaps @robpblake or @danielezonca or @r00ta know).

File: manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java [hunk 5]
```diff
@@ -186,28 +181,22 @@ public Long getProcessorsCount(String bridgeId, String customerId) {
...
+    public void deleteProcessor(String bridgeId, String processorId, String customerId) {
         Processor processor = processorDAO.findByIdBridgeIdAndCustomerId(processorId, bridgeId, customerId);
         if (processor == null) {
             throw new ItemNotFoundException(String.format("Processor with id '%s' does not exist on bridge '%s' for customer '%s'", processorId, bridgeId, customerId));
         }
-        processor.setStatus(ManagedResourceStatus.DEPROVISION);
+
+        // Processor and Connector deletion and related Work creation should always be in the same transaction
+        processorDAO.getEntityManager().merge(processor).setStatus(ManagedResourceStatus.DEPROVISION);
+        connectorService.deleteConnectorEntity(processor);
```
Comment (jomarko) [inline manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java:193]: Could you explain me a little bit the `ConnectorsService` naming. That interface seems to be tightly connected with `Processors`. Just asking, shouldn't be the names more like:
- `ProcessorsService`
- `createProcessor`
- `deleteProcessor`

Or is the naming as it is because, most Processors use some `Connectors` ?

**Review by jomarko — APPROVED** [review]
I am approving, as I have just one general question, to understand code better, however not related tho this particular PR. Thank you @manstis

**Comment by jomarko** [comment]
It was enough to re-label the PR

There is one fail, but not sure if that is related to your PR
```
Error:  [ERROR] Slack Action Processor is created and slack message should be received in the slack channel  Time elapsed: 384.835 s  <<< ERROR!
```

File: manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java (hunk 5 above)
Comment (manstis) [reply manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java:193]: Hi @jomarko I doubt there is any convention about the naming of methods in the four services, `ShardService`, `BridgeService`, `ProcessorService` and `ConnectorService`. I suspect they've been written by different people at different times. I've merely come along to make some changes. There probably is room for consistency in naming but it's arguably out of the scope of this JIRA/PR. @lucamolteni @r00ta @robpblake Would you like to impose some consistency?

Comment (jomarko) [reply manager/src/main/java/com/redhat/service/bridge/manager/ProcessorServiceImpl.java:193]: @manstis yes, I approved the PR. For sure not related to it. Just surprised that `ConnectorsService` interface methods take as argument `processor`, when we have `ProcessorService` already.
